XMLStreamException - Premature end of file

When attempting to convert XML files, we frequently run into `XMLStreamException: ParseError at [row,col]:[1,1]
Message: Premature end of file.`.

With same code and same input source and mapping this happens frequently but not always.

I created a sample project that works for me to provoke the exception (using carml version 0.4.8): 
https://github.com/mchlrch/carml-xml-issue-1

Be aware that it doesn't reliably reproduce the problem, so sometimes `XmlTestFail` may succeed and `XmlTestSuccess` may fail. It is strange and random, but if I toggle between them and try couple of times, then sooner or later the exception shows up. 

It also seems that some parts are static. So although we build a new `RdfRmlMapper` for each conversion, once the exception occurs, subsequent attempts in the same JVM will fail. I also noticed, that it doesn't work first and then fails, but fails on the first conversion attempt.

Steps to reproduce:

```
mira@mauna:~/git/carml-xml-issue-1$ java --version
openjdk 11.0.15 2022-04-19
OpenJDK Runtime Environment (build 11.0.15+10-Ubuntu-0ubuntu0.20.04.1)
OpenJDK 64-Bit Server VM (build 11.0.15+10-Ubuntu-0ubuntu0.20.04.1, mixed mode, sharing)

mira@mauna:~/git/carml-xml-issue-1$ mvn exec:java -Dexec.mainClass="XmlTestSuccess"
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.inject.internal.cglib.core.$ReflectUtils$1 (file:/usr/share/maven/lib/guice.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of com.google.inject.internal.cglib.core.$ReflectUtils$1
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
[INFO] Scanning for projects...
[INFO]
[INFO] --------------------< com.zazuko:carml-xml-issue-1 >--------------------
[INFO] Building carml-xml-issue-1 0.0.1-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- exec-maven-plugin:3.1.0:java (default-cli) @ carml-xml-issue-1 ---
SLF4J: No SLF4J providers were found.
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See https://www.slf4j.org/codes.html#noProviders for further details.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  1.408 s
[INFO] Finished at: 2023-09-06T14:26:33+02:00
[INFO] ------------------------------------------------------------------------

mira@mauna:~/git/carml-xml-issue-1$ mvn exec:java -Dexec.mainClass="XmlTestFail"
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.inject.internal.cglib.core.$ReflectUtils$1 (file:/usr/share/maven/lib/guice.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of com.google.inject.internal.cglib.core.$ReflectUtils$1
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
[INFO] Scanning for projects...
[INFO]
[INFO] --------------------< com.zazuko:carml-xml-issue-1 >--------------------
[INFO] Building carml-xml-issue-1 0.0.1-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- exec-maven-plugin:3.1.0:java (default-cli) @ carml-xml-issue-1 ---
SLF4J: No SLF4J providers were found.
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See https://www.slf4j.org/codes.html#noProviders for further details.
[WARNING]
io.carml.logicalsourceresolver.LogicalSourceResolverException: Error executing XPath expression.
    at io.carml.logicalsourceresolver.XPathResolver.xpathPathFlux (XPathResolver.java:154)
    at io.carml.logicalsourceresolver.XPathResolver.lambda$getXpathResultFlux$2 (XPathResolver.java:120)
    at reactor.core.publisher.FluxCreate.subscribe (FluxCreate.java:95)
    at reactor.core.publisher.Flux.subscribe (Flux.java:8671)
    at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap (FluxFlatMap.java:200)
    at reactor.core.publisher.FluxFlatMap.subscribeOrReturn (FluxFlatMap.java:93)
    at reactor.core.publisher.Flux.subscribe (Flux.java:8657)
    at reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext (FluxFlatMap.java:427)
    at reactor.core.publisher.FluxIterable$IterableSubscription.slowPath (FluxIterable.java:335)
    at reactor.core.publisher.FluxIterable$IterableSubscription.request (FluxIterable.java:294)
    at reactor.core.publisher.FluxFlatMap$FlatMapMain.onSubscribe (FluxFlatMap.java:371)
    at reactor.core.publisher.FluxIterable.subscribe (FluxIterable.java:201)
    at reactor.core.publisher.FluxIterable.subscribe (FluxIterable.java:83)
    at reactor.core.publisher.Flux.subscribe (Flux.java:8671)
    at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onComplete (FluxConcatArray.java:258)
    at reactor.core.publisher.FluxConcatArray.subscribe (FluxConcatArray.java:78)
    at reactor.core.publisher.Mono.subscribe (Mono.java:4485)
    at reactor.core.publisher.Mono.block (Mono.java:1733)
    at io.carml.engine.rdf.RdfRmlMapper.toModel (RdfRmlMapper.java:368)
    at io.carml.engine.rdf.RdfRmlMapper.mapToModel (RdfRmlMapper.java:351)
    at XmlTestCommon.doConversion_0 (XmlTestCommon.java:57)
    at XmlTestFail.main (XmlTestFail.java:6)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:279)
    at java.lang.Thread.run (Thread.java:829)
Caused by: javax.xml.xpath.XPathException: org.xml.sax.SAXException: javax.xml.stream.XMLStreamException: ParseError at [row,col]:[1,1]
Message: Premature end of file.
javax.xml.stream.XMLStreamException: ParseError at [row,col]:[1,1]
Message: Premature end of file.
    at jlibs.xml.sax.dog.XMLDog.sniff (XMLDog.java:192)
    at io.carml.logicalsourceresolver.XPathResolver.xpathPathFlux (XPathResolver.java:152)
    at io.carml.logicalsourceresolver.XPathResolver.lambda$getXpathResultFlux$2 (XPathResolver.java:120)
    at reactor.core.publisher.FluxCreate.subscribe (FluxCreate.java:95)
    at reactor.core.publisher.Flux.subscribe (Flux.java:8671)
    at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap (FluxFlatMap.java:200)
    at reactor.core.publisher.FluxFlatMap.subscribeOrReturn (FluxFlatMap.java:93)
    at reactor.core.publisher.Flux.subscribe (Flux.java:8657)
    at reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext (FluxFlatMap.java:427)
    at reactor.core.publisher.FluxIterable$IterableSubscription.slowPath (FluxIterable.java:335)
    at reactor.core.publisher.FluxIterable$IterableSubscription.request (FluxIterable.java:294)
    at reactor.core.publisher.FluxFlatMap$FlatMapMain.onSubscribe (FluxFlatMap.java:371)
    at reactor.core.publisher.FluxIterable.subscribe (FluxIterable.java:201)
    at reactor.core.publisher.FluxIterable.subscribe (FluxIterable.java:83)
    at reactor.core.publisher.Flux.subscribe (Flux.java:8671)
    at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onComplete (FluxConcatArray.java:258)
    at reactor.core.publisher.FluxConcatArray.subscribe (FluxConcatArray.java:78)
    at reactor.core.publisher.Mono.subscribe (Mono.java:4485)
    at reactor.core.publisher.Mono.block (Mono.java:1733)
    at io.carml.engine.rdf.RdfRmlMapper.toModel (RdfRmlMapper.java:368)
    at io.carml.engine.rdf.RdfRmlMapper.mapToModel (RdfRmlMapper.java:351)
    at XmlTestCommon.doConversion_0 (XmlTestCommon.java:57)
    at XmlTestFail.main (XmlTestFail.java:6)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:279)
    at java.lang.Thread.run (Thread.java:829)
Caused by: org.xml.sax.SAXException: javax.xml.stream.XMLStreamException: ParseError at [row,col]:[1,1]
Message: Premature end of file.
    at io.carml.logicalsourceresolver.PausableStaxXmlReader.run (PausableStaxXmlReader.java:152)
    at io.carml.logicalsourceresolver.PausableStaxXmlReader.start (PausableStaxXmlReader.java:103)
    at io.carml.logicalsourceresolver.PausableStaxXmlReader.parse (PausableStaxXmlReader.java:75)
    at jlibs.xml.sax.dog.XMLDog.sniff (XMLDog.java:189)
    at io.carml.logicalsourceresolver.XPathResolver.xpathPathFlux (XPathResolver.java:152)
    at io.carml.logicalsourceresolver.XPathResolver.lambda$getXpathResultFlux$2 (XPathResolver.java:120)
    at reactor.core.publisher.FluxCreate.subscribe (FluxCreate.java:95)
    at reactor.core.publisher.Flux.subscribe (Flux.java:8671)
    at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap (FluxFlatMap.java:200)
    at reactor.core.publisher.FluxFlatMap.subscribeOrReturn (FluxFlatMap.java:93)
    at reactor.core.publisher.Flux.subscribe (Flux.java:8657)
    at reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext (FluxFlatMap.java:427)
    at reactor.core.publisher.FluxIterable$IterableSubscription.slowPath (FluxIterable.java:335)
    at reactor.core.publisher.FluxIterable$IterableSubscription.request (FluxIterable.java:294)
    at reactor.core.publisher.FluxFlatMap$FlatMapMain.onSubscribe (FluxFlatMap.java:371)
    at reactor.core.publisher.FluxIterable.subscribe (FluxIterable.java:201)
    at reactor.core.publisher.FluxIterable.subscribe (FluxIterable.java:83)
    at reactor.core.publisher.Flux.subscribe (Flux.java:8671)
    at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onComplete (FluxConcatArray.java:258)
    at reactor.core.publisher.FluxConcatArray.subscribe (FluxConcatArray.java:78)
    at reactor.core.publisher.Mono.subscribe (Mono.java:4485)
    at reactor.core.publisher.Mono.block (Mono.java:1733)
    at io.carml.engine.rdf.RdfRmlMapper.toModel (RdfRmlMapper.java:368)
    at io.carml.engine.rdf.RdfRmlMapper.mapToModel (RdfRmlMapper.java:351)
    at XmlTestCommon.doConversion_0 (XmlTestCommon.java:57)
    at XmlTestFail.main (XmlTestFail.java:6)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:279)
    at java.lang.Thread.run (Thread.java:829)
Caused by: javax.xml.stream.XMLStreamException: ParseError at [row,col]:[1,1]
Message: Premature end of file.
    at com.sun.org.apache.xerces.internal.impl.XMLStreamReaderImpl.next (XMLStreamReaderImpl.java:652)
    at io.carml.logicalsourceresolver.PausableStaxXmlReader.run (PausableStaxXmlReader.java:150)
    at io.carml.logicalsourceresolver.PausableStaxXmlReader.start (PausableStaxXmlReader.java:103)
    at io.carml.logicalsourceresolver.PausableStaxXmlReader.parse (PausableStaxXmlReader.java:75)
    at jlibs.xml.sax.dog.XMLDog.sniff (XMLDog.java:189)
    at io.carml.logicalsourceresolver.XPathResolver.xpathPathFlux (XPathResolver.java:152)
    at io.carml.logicalsourceresolver.XPathResolver.lambda$getXpathResultFlux$2 (XPathResolver.java:120)
    at reactor.core.publisher.FluxCreate.subscribe (FluxCreate.java:95)
    at reactor.core.publisher.Flux.subscribe (Flux.java:8671)
    at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap (FluxFlatMap.java:200)
    at reactor.core.publisher.FluxFlatMap.subscribeOrReturn (FluxFlatMap.java:93)
    at reactor.core.publisher.Flux.subscribe (Flux.java:8657)
    at reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext (FluxFlatMap.java:427)
    at reactor.core.publisher.FluxIterable$IterableSubscription.slowPath (FluxIterable.java:335)
    at reactor.core.publisher.FluxIterable$IterableSubscription.request (FluxIterable.java:294)
    at reactor.core.publisher.FluxFlatMap$FlatMapMain.onSubscribe (FluxFlatMap.java:371)
    at reactor.core.publisher.FluxIterable.subscribe (FluxIterable.java:201)
    at reactor.core.publisher.FluxIterable.subscribe (FluxIterable.java:83)
    at reactor.core.publisher.Flux.subscribe (Flux.java:8671)
    at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onComplete (FluxConcatArray.java:258)
    at reactor.core.publisher.FluxConcatArray.subscribe (FluxConcatArray.java:78)
    at reactor.core.publisher.Mono.subscribe (Mono.java:4485)
    at reactor.core.publisher.Mono.block (Mono.java:1733)
    at io.carml.engine.rdf.RdfRmlMapper.toModel (RdfRmlMapper.java:368)
    at io.carml.engine.rdf.RdfRmlMapper.mapToModel (RdfRmlMapper.java:351)
    at XmlTestCommon.doConversion_0 (XmlTestCommon.java:57)
    at XmlTestFail.main (XmlTestFail.java:6)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:279)
    at java.lang.Thread.run (Thread.java:829)
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  1.544 s
[INFO] Finished at: 2023-09-06T14:26:39+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:3.1.0:java (default-cli) on project carml-xml-issue-1: An exception occurred while executing the Java class. Error executing XPath expression.: org.xml.sax.SAXException: javax.xml.stream.XMLStreamException: ParseError at [row,col]:[1,1]
[ERROR] Message: Premature end of file.
[ERROR] javax.xml.stream.XMLStreamException: ParseError at [row,col]:[1,1]
[ERROR] Message: Premature end of file.
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

--- a/carml-model/src/main/java/io/carml/model/impl/CarmlDatatypeMap.java
+++ b/carml-model/src/main/java/io/carml/model/impl/CarmlDatatypeMap.java
@@ -5,6 +5,7 @@ import io.carml.model.Resource;
 import io.carml.model.TriplesMap;
 import io.carml.vocab.Rml;
 import java.util.Set;
+import lombok.EqualsAndHashCode;
 import lombok.NoArgsConstructor;
 import lombok.experimental.SuperBuilder;
 import org.eclipse.rdf4j.model.Value;
@@ -13,6 +14,7 @@ import org.eclipse.rdf4j.model.vocabulary.RDF;
 
 @SuperBuilder
 @NoArgsConstructor
+@EqualsAndHashCode(callSuper = true)
 public class CarmlDatatypeMap extends CarmlExpressionMap implements DatatypeMap {
 
   public CarmlDatatypeMap(String id, String label, String reference, String template, Value constant,

--- a/carml-model/src/main/java/io/carml/model/impl/CarmlExpressionMap.java
+++ b/carml-model/src/main/java/io/carml/model/impl/CarmlExpressionMap.java
@@ -11,6 +11,7 @@ import io.carml.vocab.Rml;
 import io.carml.vocab.Rr;
 import java.util.Set;
 import lombok.AccessLevel;
+import lombok.EqualsAndHashCode;
 import lombok.NoArgsConstructor;
 import lombok.Setter;
 import lombok.experimental.SuperBuilder;
@@ -20,6 +21,7 @@ import org.eclipse.rdf4j.model.util.ModelBuilder;
 @SuppressWarnings("java:S1135")
 @SuperBuilder
 @NoArgsConstructor(access = AccessLevel.PACKAGE)
+@EqualsAndHashCode(callSuper = true)
 abstract class CarmlExpressionMap extends CarmlResource implements ExpressionMap {
 
   CarmlExpressionMap(String id, String label, String reference, String template, Value constant,

--- a/carml-model/src/main/java/io/carml/model/impl/CarmlJoin.java
+++ b/carml-model/src/main/java/io/carml/model/impl/CarmlJoin.java
@@ -6,6 +6,7 @@ import io.carml.rdfmapper.annotations.RdfProperty;
 import io.carml.vocab.Rdf;
 import io.carml.vocab.Rr;
 import java.util.Set;
+import lombok.EqualsAndHashCode;
 import lombok.NoArgsConstructor;
 import lombok.Setter;
 import lombok.experimental.SuperBuilder;
@@ -16,6 +17,7 @@ import org.eclipse.rdf4j.model.vocabulary.RDF;
 
 @SuperBuilder
 @NoArgsConstructor
+@EqualsAndHashCode(callSuper = false)
 public class CarmlJoin extends CarmlResource implements Join {
 
   @Setter

--- a/carml-model/src/main/java/io/carml/model/impl/CarmlLanguageMap.java
+++ b/carml-model/src/main/java/io/carml/model/impl/CarmlLanguageMap.java
@@ -4,6 +4,7 @@ import io.carml.model.LanguageMap;
 import io.carml.model.Resource;
 import io.carml.vocab.Rml;
 import java.util.Set;
+import lombok.EqualsAndHashCode;
 import lombok.NoArgsConstructor;
 import lombok.experimental.SuperBuilder;
 import org.eclipse.rdf4j.model.util.ModelBuilder;
@@ -11,6 +12,7 @@ import org.eclipse.rdf4j.model.vocabulary.RDF;
 
 @SuperBuilder
 @NoArgsConstructor
+@EqualsAndHashCode(callSuper = true)
 public class CarmlLanguageMap extends CarmlExpressionMap implements LanguageMap {
 
   @Override

--- a/carml-model/src/main/java/io/carml/model/impl/CarmlLogicalSource.java
+++ b/carml-model/src/main/java/io/carml/model/impl/CarmlLogicalSource.java
@@ -5,8 +5,8 @@ import io.carml.model.Resource;
 import io.carml.rdfmapper.annotations.RdfProperty;
 import io.carml.vocab.Rdf;
 import io.carml.vocab.Rml;
-import java.util.Objects;
 import java.util.Set;
+import lombok.EqualsAndHashCode;
 import lombok.NoArgsConstructor;
 import lombok.Setter;
 import lombok.experimental.SuperBuilder;
@@ -18,6 +18,7 @@ import org.eclipse.rdf4j.model.vocabulary.RDF;
 
 @SuperBuilder
 @NoArgsConstructor
+@EqualsAndHashCode(callSuper = false)
 public class CarmlLogicalSource extends CarmlResource implements LogicalSource {
 
   @Setter
@@ -50,27 +51,6 @@ public class CarmlLogicalSource extends CarmlResource implements LogicalSource {
   @Override
   public String toString() {
     return new ReflectionToStringBuilder(this, new MultilineRecursiveToStringStyle()).toString();
-  }
-
-  @Override
-  public int hashCode() {
-    return Objects.hash(source, iterator, referenceFormulation);
-  }
-
-  @Override
-  public boolean equals(Object obj) {
-    if (this == obj) {
-      return true;
-    }
-    if (obj == null) {
-      return false;
-    }
-    if (getClass() != obj.getClass()) {
-      return false;
-    }
-    CarmlLogicalSource other = (CarmlLogicalSource) obj;
-    return Objects.equals(source, other.source) && Objects.equals(iterator, other.iterator)
-        && Objects.equals(referenceFormulation, other.referenceFormulation);
   }
 
   @Override

--- a/carml-model/src/main/java/io/carml/model/impl/CarmlNamespace.java
+++ b/carml-model/src/main/java/io/carml/model/impl/CarmlNamespace.java
@@ -6,6 +6,7 @@ import io.carml.rdfmapper.annotations.RdfProperty;
 import io.carml.vocab.Carml;
 import io.carml.vocab.Rdf;
 import java.util.Set;
+import lombok.EqualsAndHashCode;
 import lombok.NoArgsConstructor;
 import lombok.Setter;
 import lombok.experimental.SuperBuilder;
@@ -16,6 +17,7 @@ import org.eclipse.rdf4j.model.vocabulary.RDF;
 
 @SuperBuilder
 @NoArgsConstructor
+@EqualsAndHashCode(callSuper = false)
 public class CarmlNamespace extends CarmlResource implements Namespace {
 
   @Setter

--- a/carml-model/src/main/java/io/carml/model/impl/CarmlObjectMap.java
+++ b/carml-model/src/main/java/io/carml/model/impl/CarmlObjectMap.java
@@ -10,6 +10,7 @@ import io.carml.rdfmapper.annotations.RdfType;
 import io.carml.vocab.Rdf;
 import io.carml.vocab.Rml;
 import java.util.Set;
+import lombok.EqualsAndHashCode;
 import lombok.NoArgsConstructor;
 import lombok.Setter;
 import lombok.experimental.SuperBuilder;
@@ -20,6 +21,7 @@ import org.eclipse.rdf4j.model.vocabulary.RDF;
 
 @SuperBuilder
 @NoArgsConstructor
+@EqualsAndHashCode(callSuper = true)
 public class CarmlObjectMap extends CarmlTermMap implements ObjectMap {
 
   @Setter

--- a/carml-model/src/main/java/io/carml/model/impl/CarmlPredicateMap.java
+++ b/carml-model/src/main/java/io/carml/model/impl/CarmlPredicateMap.java
@@ -5,6 +5,7 @@ import io.carml.model.PredicateMap;
 import io.carml.model.Resource;
 import io.carml.vocab.Rdf;
 import java.util.Set;
+import lombok.EqualsAndHashCode;
 import lombok.NoArgsConstructor;
 import lombok.experimental.SuperBuilder;
 import org.apache.commons.lang3.builder.MultilineRecursiveToStringStyle;
@@ -14,6 +15,7 @@ import org.eclipse.rdf4j.model.vocabulary.RDF;
 
 @SuperBuilder
 @NoArgsConstructor
+@EqualsAndHashCode(callSuper = true)
 public class CarmlPredicateMap extends CarmlTermMap implements PredicateMap {
 
   @Override

--- a/carml-model/src/main/java/io/carml/model/impl/CarmlPredicateObjectMap.java
+++ b/carml-model/src/main/java/io/carml/model/impl/CarmlPredicateObjectMap.java
@@ -12,6 +12,7 @@ import io.carml.rdfmapper.annotations.RdfTypeDecider;
 import io.carml.vocab.Rdf;
 import io.carml.vocab.Rr;
 import java.util.Set;
+import lombok.EqualsAndHashCode;
 import lombok.NoArgsConstructor;
 import lombok.Setter;
 import lombok.Singular;
@@ -23,6 +24,7 @@ import org.eclipse.rdf4j.model.vocabulary.RDF;
 
 @SuperBuilder
 @NoArgsConstructor
+@EqualsAndHashCode(callSuper = false)
 public class CarmlPredicateObjectMap extends CarmlResource implements PredicateObjectMap {
 
   @Singular

--- a/carml-model/src/main/java/io/carml/model/impl/CarmlRefObjectMap.java
+++ b/carml-model/src/main/java/io/carml/model/impl/CarmlRefObjectMap.java
@@ -11,6 +11,7 @@ import io.carml.vocab.Carml;
 import io.carml.vocab.Rdf;
 import io.carml.vocab.Rr;
 import java.util.Set;
+import lombok.EqualsAndHashCode;
 import lombok.NoArgsConstructor;
 import lombok.Setter;
 import lombok.Singular;
@@ -22,6 +23,7 @@ import org.eclipse.rdf4j.model.vocabulary.RDF;
 
 @SuperBuilder
 @NoArgsConstructor
+@EqualsAndHashCode(callSuper = false)
 public class CarmlRefObjectMap extends CarmlResource implements RefObjectMap {
 
   @Setter

--- a/carml-model/src/main/java/io/carml/model/impl/CarmlResource.java
+++ b/carml-model/src/main/java/io/carml/model/impl/CarmlResource.java
@@ -90,13 +90,14 @@ public abstract class CarmlResource implements Resource {
   private Model asRdfInternal(Resource resource, Set<Resource> processed) {
     processed.add(resource);
     ModelBuilder builder = new ModelBuilder();
-    addTriples(builder);
+    resource.addTriples(builder);
     if (label != null) {
       builder.add(RDFS.LABEL, label);
     }
     Model model = builder.build();
 
-    Model nestedModel = getReferencedResources().stream()
+    Model nestedModel = resource.getReferencedResources()
+        .stream()
         .filter(refResource -> !processed.contains(refResource))
         .flatMap(refResource -> asRdfInternal(refResource, processed).stream())
         .collect(ModelCollector.toModel());

--- a/carml-model/src/main/java/io/carml/model/impl/CarmlSubjectMap.java
+++ b/carml-model/src/main/java/io/carml/model/impl/CarmlSubjectMap.java
@@ -9,6 +9,7 @@ import io.carml.rdfmapper.annotations.RdfType;
 import io.carml.vocab.Rdf;
 import io.carml.vocab.Rr;
 import java.util.Set;
+import lombok.EqualsAndHashCode;
 import lombok.NoArgsConstructor;
 import lombok.Singular;
 import lombok.experimental.SuperBuilder;
@@ -20,6 +21,7 @@ import org.eclipse.rdf4j.model.vocabulary.RDF;
 
 @SuperBuilder
 @NoArgsConstructor
+@EqualsAndHashCode(callSuper = true)
 public class CarmlSubjectMap extends CarmlTermMap implements SubjectMap {
 
   @Singular("clazz")

--- a/carml-model/src/main/java/io/carml/model/impl/CarmlTermMap.java
+++ b/carml-model/src/main/java/io/carml/model/impl/CarmlTermMap.java
@@ -6,6 +6,7 @@ import io.carml.rdfmapper.annotations.RdfProperty;
 import io.carml.vocab.Rdf;
 import io.carml.vocab.Rr;
 import lombok.AccessLevel;
+import lombok.EqualsAndHashCode;
 import lombok.NoArgsConstructor;
 import lombok.Setter;
 import lombok.experimental.SuperBuilder;
@@ -13,6 +14,7 @@ import org.eclipse.rdf4j.model.util.ModelBuilder;
 
 @SuperBuilder
 @NoArgsConstructor(access = AccessLevel.PACKAGE)
+@EqualsAndHashCode(callSuper = true)
 abstract class CarmlTermMap extends CarmlExpressionMap implements TermMap {
 
   @Setter

--- a/carml-model/src/main/java/io/carml/model/impl/CarmlTriplesMap.java
+++ b/carml-model/src/main/java/io/carml/model/impl/CarmlTriplesMap.java
@@ -12,6 +12,7 @@ import io.carml.vocab.Rdf;
 import io.carml.vocab.Rml;
 import io.carml.vocab.Rr;
 import java.util.Set;
+import lombok.EqualsAndHashCode;
 import lombok.NoArgsConstructor;
 import lombok.Setter;
 import lombok.Singular;
@@ -24,6 +25,7 @@ import org.eclipse.rdf4j.model.vocabulary.RDF;
 
 @SuperBuilder
 @NoArgsConstructor
+@EqualsAndHashCode(callSuper = false)
 public class CarmlTriplesMap extends CarmlResource implements TriplesMap {
 
   @Setter

--- a/carml-rdf-mapper/src/main/java/io/carml/rdfmapper/impl/CarmlMapper.java
+++ b/carml-rdf-mapper/src/main/java/io/carml/rdfmapper/impl/CarmlMapper.java
@@ -163,6 +163,7 @@ public class CarmlMapper implements Mapper, MappingCache {
             throw new IllegalStateException(
                 String.format("No combiner specified for non-void multi delegate method %S", method));
           }
+          delegates.forEach(delegate -> singleDelegateMethodInvocation(delegate, method, args));
           return null;
         });
   }
@@ -191,8 +192,6 @@ public class CarmlMapper implements Mapper, MappingCache {
           String.format("error trying to invoke method [%s] on delegate [%s]", method, delegate), e);
     }
   }
-
-
 
   private Stream<Method> gatherMethods(Class<?> clazz) {
     return concat(stream(clazz.getDeclaredMethods()),

